app/models/product_import/product_importer.rb

```diff
@@ -267,11 +259,11 @@ def delete_uploaded_file
       File.delete(@file)
     end
 
-    def build_entries_from_rows(rows)
+    def build_entries_from_rows(rows, offset = 0)
       rows.each_with_index.inject([]) do |entries, (row, i)|
         row_data = Hash[[headers, row].transpose]
         entry = SpreadsheetEntry.new(row_data)
-        entry.line_number = i + 2
+        entry.line_number = offset + i + 1
```

Comment:
I leave 2 questions:
1- here you are changing the previous behaviour called from build_entries right? it was i + 2 and now after this it's  offset (0) + i + 1
2- in this new version you skipped the logic in the break statement, right?    break if @sheet.last_row == line_number

and 2 recommendations
1- variable offset, I'd name it line_number_offset, much easier to understand
2- variable start_line and end_line  indicate a line but it's actually a number, better name is start_line_number and end_line_number

Reply:
Noting again that the batch size is 50. The `break if @sheet.last_row == line_number` was to make sure that, if there were 52 rows, the block would no longer need to be executed for rows 53 to 100.

I moved this logic out of the `build_entries_from_rows` method so we can reuse the method. Skipping the non-existent rows already happens here: `rows[(start_line - 1)..(end_line - 1)]`. This expression will end at the last element in `rows`.

```
> [1, 2, 3, 4, 5][0..3]
=> [1, 2, 3, 4]
> [1, 2, 3, 4, 5][0..4]
=> [1, 2, 3, 4, 5]
> [1, 2, 3, 4, 5][0..6]
=> [1, 2, 3, 4, 5]
```

Reply:
@luisramos0 About the Point 1, thanks. I updated the commit here: https://github.com/openfoodfoundation/openfoodnetwork/pull/3908/commits/38519b2baeabab27fa9c7097d9130d461f0da5a7 I renamed the variable names and added a comment. I was getting confused among line number (with header, start at 1), row index (with header, start at 0), and data row index (no header, start at 0).

Reply:
[1, 2, 3, 4, 5][0..6] = [1, 2, 3, 4, 5]
:-O :+1: